In [1]:
print("hello world")

hello world


Pandas is an open-source python library that is used for data manipulation and analysis. It provides many functions and methods to speed up the data analysis process. Pandas is built on top of the NumPy package, hence it takes a lot of basic inspiration from it. The two primary data structures are Series which is 1 dimensional and DataFrame which is 2 dimensional.

It is one of the most important and useful tools in the arsenal of a Data Scientist and a Data Analyst.

In [2]:
# pip install pandas


In [4]:
import pandas as pd
import os

In [5]:
# def getFilePath(filename):
#     currentDir  = os.getcwd()
#     fullPath = os.path.join(currentDir, filename)
#     return fullPath

# df = pd.read_csv(getFilePath('police.csv'))
# print(df.head())

    stop_date stop_time  county_name driver_gender  driver_age_raw  \
0  2005-01-02     01:55          NaN             M          1985.0   
1  2005-01-18     08:15          NaN             M          1965.0   
2  2005-01-23     23:15          NaN             M          1972.0   
3  2005-02-20     17:15          NaN             M          1986.0   
4  2005-03-14     10:00          NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 

In [6]:
df = pd.read_csv("police.csv")

head()

Then, we will use pandas “head()” function to display the top 5 rows from our data set. Note – We can provide the no. of rows that we want to display by providing the count as a parameter to the “head()” function, i.e. (df.head(10) – this will now display 10 rows from our dataset).

In [7]:
df.head()

stop_date stop_time  county_name driver_gender  driver_age_raw  \
0  2005-01-02     01:55          NaN             M          1985.0   
1  2005-01-18     08:15          NaN             M          1965.0   
2  2005-01-23     23:15          NaN             M          1972.0   
3  2005-02-20     17:15          NaN             M          1986.0   
4  2005-03-14     10:00          NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

There is a “tail()” method as well which will show the last 5 rows from our data set.

In [8]:
df.tail()

stop_date stop_time  county_name driver_gender  driver_age_raw  \
91736  2015-12-31     20:27          NaN             M          1986.0   
91737  2015-12-31     20:35          NaN             F          1982.0   
91738  2015-12-31     20:45          NaN             M          1992.0   
91739  2015-12-31     21:42          NaN             M          1993.0   
91740  2015-12-31     22:46          NaN             M          1959.0   

       driver_age driver_race                   violation_raw  \
91736        29.0       White                        Speeding   
91737        33.0       White  Equipment/Inspection Violation   
91738        23.0       White         Other Traffic Violation   
91739        22.0       White                        Speeding   
91740        56.0    Hispanic                        Speeding   

              violation  search_conducted search_type stop_outcome  \
91736          Speeding             False         NaN      Warning   
91737         Equipment             False         NaN      Warning   
91738  Moving violation             False         NaN      Warning   
91739          Speeding             False         NaN     Citation   
91740          Speeding             False         NaN     Citation   

      is_arrested stop_duration  drugs_related_stop  
91736       False      0-15 Min               False  
91737       False      0-15 Min               False  
91738       False      0-15 Min               False  
91739       False      0-15 Min               False  
91740       False      0-15 Min               False

shape

Now, if we want to see the dimension of our dataset, we can use “shape”, which will show the dimensions in (No. of rows, No. of columns) format.

In [10]:
df.shape

(91741, 15)

info()

Now, if we want to know some more information about our dataset, we can use the “info()” function of pandas. It displays various information about our data such as the column names, no. of non-null values in each column(feature), type of each column, memory usage, etc.

In [11]:
df.info

<bound method DataFrame.info of         stop_date stop_time  county_name driver_gender  driver_age_raw  \
0      2005-01-02     01:55          NaN             M          1985.0   
1      2005-01-18     08:15          NaN             M          1965.0   
2      2005-01-23     23:15          NaN             M          1972.0   
3      2005-02-20     17:15          NaN             M          1986.0   
4      2005-03-14     10:00          NaN             F          1984.0   
...           ...       ...          ...           ...             ...   
91736  2015-12-31     20:27          NaN             M          1986.0   
91737  2015-12-31     20:35          NaN             F          1982.0   
91738  2015-12-31     20:45          NaN             M          1992.0   
91739  2015-12-31     21:42          NaN             M          1993.0   
91740  2015-12-31     22:46          NaN             M          1959.0   

       driver_age driver_race                   violation_raw  \
0            2

to_datetime()

So, while reading a CSV file, the DateTime objects in the file are read as string objects and therefore, it’s a little difficult to perform DateTime operations like time difference on a string. So, this is where the pandas “to_datetime()” method comes into play. You can provide various formats as per your requirement.

In [12]:
import datetime as dt

In [13]:
df['stop_date'] = pd.to_datetime(df['stop_date'])

In [15]:
# df['stop_time'] = pd.to_datetime(df['stop_time'], format='%H:%M:').dt.time

In [16]:
df['stop_time'] = pd.to_datetime(df['stop_time'], format='%H:%M').dt.time


In [17]:
df.info

<bound method DataFrame.info of        stop_date stop_time  county_name driver_gender  driver_age_raw  \
0     2005-01-02  01:55:00          NaN             M          1985.0   
1     2005-01-18  08:15:00          NaN             M          1965.0   
2     2005-01-23  23:15:00          NaN             M          1972.0   
3     2005-02-20  17:15:00          NaN             M          1986.0   
4     2005-03-14  10:00:00          NaN             F          1984.0   
...          ...       ...          ...           ...             ...   
91736 2015-12-31  20:27:00          NaN             M          1986.0   
91737 2015-12-31  20:35:00          NaN             F          1982.0   
91738 2015-12-31  20:45:00          NaN             M          1992.0   
91739 2015-12-31  21:42:00          NaN             M          1993.0   
91740 2015-12-31  22:46:00          NaN             M          1959.0   

       driver_age driver_race                   violation_raw  \
0            20.0       Wh

isnull()

Using “isnull()” and “sum()” functions, we can find that the no. of null values in a DataFrame for every feature.

In [18]:
df.isnull().sum()

stop_date                 0
stop_time                 0
county_name           91741
driver_gender          5335
driver_age_raw         5327
driver_age             5621
driver_race            5333
violation_raw          5333
violation              5333
search_conducted          0
search_type           88545
stop_outcome           5333
is_arrested            5333
stop_duration          5333
drugs_related_stop        0
dtype: int64

drop()

Now, as we can see that “county_name” column is completely empty, so it will not provide any information which is beneficial to us. Hence, we will drop that particular column using the pandas “drop()” function. Note – We provide “inplace=True’ to modify the current DataFrame.

In [20]:
df.head(1)

stop_date stop_time  county_name driver_gender  driver_age_raw  driver_age  \
0 2005-01-02  01:55:00          NaN             M          1985.0        20.0   

  driver_race violation_raw violation  search_conducted search_type  \
0       White      Speeding  Speeding             False         NaN   

  stop_outcome is_arrested stop_duration  drugs_related_stop  
0     Citation       False      0-15 Min               False

In [23]:
df.drop(['county_name'], inplace=True, axis=1)

In [25]:
df.head(1)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0 2005-01-02  01:55:00             M          1985.0        20.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
0      Speeding  Speeding             False         NaN     Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False

### describe()

Now, using the “describe()” function, we can get various information about the numerical columns in our DataFrame, such as total count of rows, mean, the minimum value, the maximum value, and the spread of the values in the particular feature.

In [1]:
df.describe()

NameError: name 'df' is not defined

### value_counts()

“value_counts()” function is used to identify the different categories in a feature as well as the count of values per category.

In [27]:
df['search_type'].values

array([nan, nan, nan, ..., nan, nan, nan], dtype=object)

In [28]:
df['search_type'].value_counts

<bound method IndexOpsMixin.value_counts of 0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
91736    NaN
91737    NaN
91738    NaN
91739    NaN
91740    NaN
Name: search_type, Length: 91741, dtype: object>

### fillna()

Now, as we know that, by using the “isnull()” and “sum()” functions, we can check if our data has any missing values or not. So, now we can see that this “driver_gender” feature has 5335 nan or missing values. So, we will fill up the missing values using the mode(a value that appears most frequently in a data set) of this particular feature using the “fillna()” function.

In [29]:
df['driver_gender'].isnull().sum()

5335

In [30]:
df['driver_gender'].fillna(df['driver_gender'].mode, inplace=True)

In [31]:
df['driver_gender'].isnull().sum()

0

Note – This way of filling the missing or nan values is not the most effective way and there are other much more efficient ways for imputation. There is a lot of thought which goes behind when we are working with missing values but, this is just for explaining the functionality of the fillna() function.

sample()

We can use the “sample()” function which allows us to choose random values from our data frame. We can pass it the no. of rows that we want to fetch as a parameter.

In [33]:
df.sample(2)

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
35995 2009-07-08  08:48:00             F          1959.0        50.0   
51731 2011-08-16  09:48:00             M          1966.0        45.0   

      driver_race            violation_raw         violation  \
35995       White                 Speeding          Speeding   
51731       Black  Other Traffic Violation  Moving violation   

       search_conducted search_type stop_outcome is_arrested stop_duration  \
35995             False         NaN     Citation       False      0-15 Min   
51731             False         NaN      Warning       False      0-15 Min   

       drugs_related_stop  
35995               False  
51731               False

nunique()

We can use the “nunique()” function to find the no. of unique values in our series or data frame. Generally, it is used in the case of categorical features to identify the no. of categories in a particular feature.

In [34]:
df['driver_race'].nunique()

5

columns

As the name suggests, “columns” get the name of all the features/columns in our data frame.

In [35]:
df.columns

Index(['stop_date', 'stop_time', 'driver_gender', 'driver_age_raw',
       'driver_age', 'driver_race', 'violation_raw', 'violation',
       'search_conducted', 'search_type', 'stop_outcome', 'is_arrested',
       'stop_duration', 'drugs_related_stop'],
      dtype='object')

nsmallest() & nlargest()

So, as the name suggests, “nsmallest() & nlargest()” functions are used to obtain “n” no. of rows from our dataset which are lowest or highest respectively.

In [37]:
df.nsmallest(4,'driver_age')

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
10500 2006-09-30  01:30:00             M          1991.0        15.0   
17771 2007-06-11  12:30:00             M          1992.0        15.0   
18357 2007-07-04  00:40:00             F          1992.0        15.0   
25294 2008-04-21  11:19:00             M          1993.0        15.0   

      driver_race            violation_raw         violation  \
10500       Black  Other Traffic Violation  Moving violation   
17771       White                 Speeding          Speeding   
18357       White  Other Traffic Violation  Moving violation   
25294    Hispanic  Other Traffic Violation  Moving violation   

       search_conducted         search_type   stop_outcome is_arrested  \
10500             False                 NaN  Arrest Driver        True   
17771             False                 NaN       Citation       False   
18357             False                 NaN  Arrest Driver        True   
25294              True  Incident to Arrest  Arrest Driver        True   

      stop_duration  drugs_related_stop  
10500       30+ Min               False  
17771      0-15 Min               False  
18357      0-15 Min               False  
25294       30+ Min               False

# groupby()

The “groupby()” function is very useful in data analysis as it allows us to unveil the underlying relationships among different variables. And then we can apply Aggregations as well on the groups with the “agg()” function and pass it with various aggregation operations such as mean, size, sum, std etc.

In [39]:
import numpy as np
# df.groupby('driver_race').agg(np.mean)

In [43]:
# # Group by 'driver_race' and calculate the mean for each group
# df_grouped_mean = df.groupby('driver_race').agg(np.mean)

# # Display the resulting DataFrame
# print(df_grouped_mean)

In [58]:
# df.head(3)

In [56]:
# # numeric_columns = ['column1', 'column2']
# numeric_columns = [ 'driver_race']


In [57]:
# df[numeric_columns] = df[numeric_columns].apply(pd.to_numeric)


In [59]:
df.head(1)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0 2005-01-02  01:55:00             M          1985.0        20.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
0      Speeding  Speeding             False         NaN     Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False

### get_group()

We can use the “get_group()” function to select a specific group.

In [62]:
# df.groupby('driver_race').get_group('Asian')
df.groupby('driver_race').get_group('Asian').head(1)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
10 2005-07-13  16:20:00             M          1979.0        26.0       Asian   

   violation_raw violation  search_conducted search_type stop_outcome  \
10      Speeding  Speeding             False         NaN     Citation   

   is_arrested stop_duration  drugs_related_stop  
10       False      0-15 Min               False

Note – We can combine various methods of pandas as per our requirement for a better understanding of data as shown below:

In [63]:
df.groupby('driver_race').driver_age.describe()

count       mean        std   min   25%   50%   75%   max
driver_race                                                             
Asian         2253.0  33.579672  11.688393  17.0  24.0  31.0  41.0  82.0
Black        12197.0  33.110519  11.493603  15.0  24.0  30.0  40.0  85.0
Hispanic      9477.0  31.565369  10.561938  15.0  23.0  29.0  38.0  76.0
Other          239.0  34.548117  11.607951  18.0  25.0  32.0  41.5  74.0
White        61949.0  34.576958  13.252149  15.0  23.0  31.0  44.0  99.0

# loc() and iloc()

loc() and iloc() methods are used in slicing data from the pandas DataFrame which helps in filtering the data according to some given condition.

### loc – select by labels

### iloc – select by positions

In [64]:
df.loc[:5,['driver_age', 'violation']]

driver_age  violation
0        20.0   Speeding
1        40.0   Speeding
2        33.0   Speeding
3        19.0      Other
4        21.0   Speeding
5        23.0  Equipment

In [66]:
df.loc[(df.driver_age < 16) &( df.violation == 'Speeding')]

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
17771 2007-06-11  12:30:00             M          1992.0        15.0   

      driver_race violation_raw violation  search_conducted search_type  \
17771       White      Speeding  Speeding             False         NaN   

      stop_outcome is_arrested stop_duration  drugs_related_stop  
17771     Citation       False      0-15 Min               False

### iloc()
 slices the data frame in the specified rows and column range.

In [67]:
df.iloc[:5, :5]

stop_date stop_time driver_gender  driver_age_raw  driver_age
0 2005-01-02  01:55:00             M          1985.0        20.0
1 2005-01-18  08:15:00             M          1965.0        40.0
2 2005-01-23  23:15:00             M          1972.0        33.0
3 2005-02-20  17:15:00             M          1986.0        19.0
4 2005-03-14  10:00:00             F          1984.0        21.0

In [68]:
df.head(1)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0 2005-01-02  01:55:00             M          1985.0        20.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
0      Speeding  Speeding             False         NaN     Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False

In [69]:
df.columns

Index(['stop_date', 'stop_time', 'driver_gender', 'driver_age_raw',
       'driver_age', 'driver_race', 'violation_raw', 'violation',
       'search_conducted', 'search_type', 'stop_outcome', 'is_arrested',
       'stop_duration', 'drugs_related_stop'],
      dtype='object')

# Sorting

We can sort our DataFrame by index or values with Pandas “sort_index()” and “sort_values()” functions. Below is the implementation for sort by values:

In [70]:
df[['stop_date','driver_age', 'driver_race', 'violation', 'stop_outcome']].sort_values('driver_age')

stop_date  driver_age driver_race         violation   stop_outcome
18357 2007-07-04        15.0       White  Moving violation  Arrest Driver
17771 2007-06-11        15.0       White          Speeding       Citation
45988 2010-11-10        15.0       White  Moving violation       Citation
10500 2006-09-30        15.0       Black  Moving violation  Arrest Driver
25294 2008-04-21        15.0    Hispanic  Moving violation  Arrest Driver
...          ...         ...         ...               ...            ...
91637 2015-12-27         NaN         NaN               NaN            NaN
91660 2015-12-28         NaN         NaN               NaN            NaN
91674 2015-12-28         NaN         NaN               NaN            NaN
91710 2015-12-30         NaN         NaN               NaN            NaN
91713 2015-12-30         NaN         NaN               NaN            NaN

[91741 rows x 5 columns]

In [71]:
df[['stop_date','driver_age', 'driver_race', 'violation', 'stop_outcome']].sort_values('driver_age').head(1)

stop_date  driver_age driver_race         violation   stop_outcome
18357 2007-07-04        15.0       White  Moving violation  Arrest Driver

### Query

We can use the Pandas query() function to filter our data frame as per our conditions or requirements as shown below:

In [72]:
df.query('45 < driver_age <50').head()

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
19 2005-07-24  20:10:00             F          1958.0        47.0       White   
30 2005-09-26  12:09:00             M          1959.0        46.0       Black   
46 2005-10-01  08:40:00             M          1959.0        46.0       White   
48 2005-10-01  09:20:00             M          1957.0        48.0       White   
57 2005-10-01  18:10:00             M          1958.0        47.0       White   

                     violation_raw         violation  search_conducted  \
19                        Speeding          Speeding             False   
30         Other Traffic Violation  Moving violation             False   
46  Equipment/Inspection Violation         Equipment             False   
48                        Speeding          Speeding             False   
57                        Speeding          Speeding             False   

   search_type stop_outcome is_arrested stop_duration  drugs_related_stop  
19         NaN     Citation       False      0-15 Min               False  
30         NaN     Citation       False       30+ Min               False  
46         NaN     Citation       False      0-15 Min               False  
48         NaN     Citation       False      0-15 Min               False  
57         NaN     Citation       False      0-15 Min               False

In [73]:
df.columns

Index(['stop_date', 'stop_time', 'driver_gender', 'driver_age_raw',
       'driver_age', 'driver_race', 'violation_raw', 'violation',
       'search_conducted', 'search_type', 'stop_outcome', 'is_arrested',
       'stop_duration', 'drugs_related_stop'],
      dtype='object')

In [75]:
df.query('driver_age ==19 and driver_gender =="F"')

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
7     2005-06-06  13:20:00             F          1986.0        19.0   
36    2005-09-30  09:30:00             F          1986.0        19.0   
66    2005-10-01  23:10:00             F          1986.0        19.0   
204   2005-10-07  07:10:00             F          1986.0        19.0   
267   2005-10-09  21:10:00             F          1986.0        19.0   
...          ...       ...           ...             ...         ...   
91427 2015-12-14  13:54:00             F          1996.0        19.0   
91432 2015-12-14  16:31:00             F          1996.0        19.0   
91576 2015-12-22  01:25:00             F          1996.0        19.0   
91672 2015-12-28  11:05:00             F          1996.0        19.0   
91730 2015-12-31  15:36:00             F          1996.0        19.0   

      driver_race                   violation_raw  violation  \
7           White                        Speeding   Speeding   
36          White                        Speeding   Speeding   
66          White                        Speeding   Speeding   
204         White                        Speeding   Speeding   
267         White  Equipment/Inspection Violation  Equipment   
...           ...                             ...        ...   
91427    Hispanic                        Speeding   Speeding   
91432       White                        Speeding   Speeding   
91576       White  Equipment/Inspection Violation  Equipment   
91672       White                             APB      Other   
91730    Hispanic  Equipment/Inspection Violation  Equipment   

       search_conducted         search_type stop_outcome is_arrested  \
7                 False                 NaN     Citation       False   
36                False                 NaN     Citation       False   
66                False                 NaN     Citation       False   
204               False                 NaN     Citation       False   
267               False                 NaN          N/D       False   
...                 ...                 ...          ...         ...   
91427             False                 NaN      Warning       False   
91432             False                 NaN     Citation       False   
91576             False                 NaN     Citation       False   
91672              True  Incident to Arrest     Citation       False   
91730             False                 NaN      Warning       False   

      stop_duration  drugs_related_stop  
7          0-15 Min               False  
36         0-15 Min               False  
66         0-15 Min               False  
204        0-15 Min               False  
267        0-15 Min               False  
...             ...                 ...  
91427      0-15 Min               False  
91432      0-15 Min               False  
91576      0-15 Min               False  
91672     16-30 Min                True  
91730      0-15 Min               False  

[977 rows x 14 columns]

In [76]:
df.query('driver_age ==19 and driver_gender =="F"').head(1)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
7 2005-06-06  13:20:00             F          1986.0        19.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
7      Speeding  Speeding             False         NaN     Citation   

  is_arrested stop_duration  drugs_related_stop  
7       False      0-15 Min               False

### set_index()

So, we can use Pandas “set_index()” function to set any of your columns as the index.

In [78]:
df.set_index('stop_date').head(2)

stop_time driver_gender  driver_age_raw  driver_age driver_race  \
stop_date                                                                    
2005-01-02  01:55:00             M          1985.0        20.0       White   
2005-01-18  08:15:00             M          1965.0        40.0       White   

           violation_raw violation  search_conducted search_type stop_outcome  \
stop_date                                                                       
2005-01-02      Speeding  Speeding             False         NaN     Citation   
2005-01-18      Speeding  Speeding             False         NaN     Citation   

           is_arrested stop_duration  drugs_related_stop  
stop_date                                                 
2005-01-02       False      0-15 Min               False  
2005-01-18       False      0-15 Min               False

### duplicated()

We can use the “duplicated()” function to find all the duplicate rows in our dataset. And, then we can remove duplicate values using the drop_duplicates() function, as having too many duplicate values will affect the accuracy of our model at the later stage.

In [79]:
print(df.duplicated().sum())

479


In [80]:
df.drop_duplicates(inplace=True)

In [81]:
print(df.duplicated().sum())

0


#### get_dummies()

Pandas “get_dummies()” method is used to convert the categorical features of the data into dummy variables or indicator variables.

We usually do this conversion because some machine learning models don’t work well with categorical values such as Random Forrest, but we shouldn’t be using this if we have too many categories in our feature as it will create that many new features in our data frame, which will have an effect on the performance of our model.

In [82]:
newdf = pd.get_dummies(df.driver_race, prefix="driver_race_").iloc[:,0:]

In [83]:
newdf.head()

driver_race__Asian  driver_race__Black  driver_race__Hispanic  \
0               False               False                  False   
1               False               False                  False   
2               False               False                  False   
3               False               False                  False   
4               False               False                  False   

   driver_race__Other  driver_race__White  
0               False                True  
1               False                True  
2               False                True  
3               False                True  
4               False                True

#### select_dtypes()

We can separate the numerical and categorical features from our data frame and create new ones by using the “select_dtypes()” function and include “np.number” to select numerical columns whereas include “objects” for categorical columns.

In [85]:
numerical_df = df.select_dtypes(include=[np.number])

In [87]:
numerical_df.head(1)

driver_age_raw  driver_age
0          1985.0        20.0

In [88]:
categorical_df =df.select_dtypes(include='object')

In [90]:
categorical_df.head(1)

stop_time driver_gender driver_race violation_raw violation search_type  \
0  01:55:00             M       White      Speeding  Speeding         NaN   

  stop_outcome is_arrested stop_duration  
0     Citation       False      0-15 Min

#### concat()

We can perform concatenation of pandas object into a DataFrame output along a particular axis with optional set logic such as union and intersection using concat() method.

By default, axis=0, i.e. row-wise concatenation, so if we set axis=1, column-wise concatenation will be performed.

In [94]:
comb_df = pd.concat([numerical_df, categorical_df], axis = 1)

In [96]:
comb_df.head(2)

driver_age_raw  driver_age stop_time driver_gender driver_race  \
0          1985.0        20.0  01:55:00             M       White   
1          1965.0        40.0  08:15:00             M       White   

  violation_raw violation search_type stop_outcome is_arrested stop_duration  
0      Speeding  Speeding         NaN     Citation       False      0-15 Min  
1      Speeding  Speeding         NaN     Citation       False      0-15 Min

### apply()

Suppose we create our own custom function and we want to use that function in our data frame. This is where the Pandas “apply()” function comes into play. It allows us to apply a custom function to every element of a particular Series.

So, here we have created our own custom function currentAge() which returns the current age of the person by subtracting their date of birth from the current year(2021). And, after that we can use this function inside the “apply()” function.

In [97]:
def currentAge(dob):
    return (2022 - dob)

In [98]:
df['driver_current_age'] = df['driver_age_raw'].apply(currentAge)

In [99]:
df['driver_current_age'].head(2)

0    37.0
1    57.0
Name: driver_current_age, dtype: float64

#### qcut() and cut()

So, when we have to deal with continuous numeric data, it is often helpful to bin them into multiple buckets and then carry on with the further analysis of the data. Pandas provide two methods which are qcut() and cut(), which helps us to convert continuous data to a set of discrete buckets.

#### qcut()
 method ensures a more even distribution of the values inside each bin, so we can say it’s a better sampling. We just pass the no. of bins and then Pandas does the behind the scene job to decide how wide to make each bin.

#### cut()
 method is used to specifically define the bin edges and hence the distribution of values is not even across all the bins. There might be a situation when there is no item inside a particular bin, so we should be careful about that.

In [100]:
df['driver_age_qcut_bins'] = pd.qcut(df["driver_age"], q = 5)

In [101]:
df['driver_age_qcut_bins'].head()

0    (14.999, 22.0]
1      (35.0, 46.0]
2      (27.0, 35.0]
3    (14.999, 22.0]
4    (14.999, 22.0]
Name: driver_age_qcut_bins, dtype: category
Categories (5, interval[float64, right]): [(14.999, 22.0] < (22.0, 27.0] < (27.0, 35.0] < (35.0, 46.0] < (46.0, 99.0]]

In [102]:
df['driver_age_qcut_bins'].value_counts()

driver_age_qcut_bins
(27.0, 35.0]      17746
(22.0, 27.0]      17630
(14.999, 22.0]    17516
(35.0, 46.0]      16812
(46.0, 99.0]      16148
Name: count, dtype: int64

In [104]:
df['driver_age_cut_bins'] = pd.cut(df['driver_age'], bins=3, labels=  ["1-21", "22-60", "61-100"])

In [105]:
df['driver_age_cut_bins'].head(2)

0    1-21
1    1-21
Name: driver_age_cut_bins, dtype: category
Categories (3, object): ['1-21' < '22-60' < '61-100']

In [106]:
df['driver_age_cut_bins'].value_counts()

driver_age_cut_bins
1-21      65535
22-60     19875
61-100      442
Name: count, dtype: int64

### to_csv()

Now, we can save our DataFrame in a CSV file using the pandas “to_csv()” function. As we don’t want to store the preceding indices of each row, hence, we will set index=False.

In [110]:
comb_df.to_csv('out/output.csv', index = False)